respy/method_of_simulated_moments.py

```diff
@@ -0,0 +1,361 @@
+    flat_simulated_moments : pandas.DataFrame
+        Vector of simulated_moments, index matches index of empirical_moments.
+
+    """
+    empirical_moments = copy.deepcopy(empirical_moments)
+
+    empirical_moments = _harmonize_input(empirical_moments)
+    calc_moments = _harmonize_input(calc_moments)
+    replace_nans = _harmonize_input(replace_nans)
+
+    simulate = get_simulate_func(params, options)
+    df_sim = simulate(params)
+    simulated_moments = _calculate_moments(df_sim, calc_moments)
+
+    simulated_moments = _harmonize_input(simulated_moments)
+
+    simulated_moments = _reindex_simulated_moments(empirical_moments, simulated_moments)
+
+    simulated_moments = _replace_missing_values(simulated_moments, replace_nans)
+
+    flat_empirical_moments = _flatten_index(empirical_moments)
+    flat_simulated_moments = _flatten_index(simulated_moments)
+
+    return flat_empirical_moments, flat_simulated_moments
+
+
+def _harmonize_input(data):
+    """Harmonize different types of inputs by turning all inputs into lists.
+
+    - pandas.DataFrames/Series and callable functions will turn into a list containing a
+      single item (i.e. the input).
+    - Dictionaries will be sorted according to keys and then turn into a list containing
+      the dictionary entries.
+
+    """
+    # Convert single DataFrames, Series or function into list containing one item.
+    if isinstance(data, (pd.DataFrame, pd.Series)) or callable(data):
+        data = [data]
+
+    # Sort dictionary according to keys and turn into list.
+    elif isinstance(data, dict):
+        data = [data[key] for key in sorted(data)]
+
+    elif isinstance(data, list):
+        pass
+
+    else:
+        raise TypeError(
+            "Function only accepts lists, dictionaries, functions, Series and "
+            "DataFrames as inputs."
+        )
+
+    return data
+
+
+def _calculate_moments(data, calc_moments):
+    """Calculate the moments."""
+    moments = []
+    for func in calc_moments:
+        moments.append(func(data))
+
+    return moments
+
+
+def _reindex_simulated_moments(empirical_moments, simulated_moments):
+    """Reindex data in simulated_moments to match index of empirical_moments."""
+    simulated_moments_reindexed = []
+
+    for emp_mom, sim_mom in zip(empirical_moments, simulated_moments):
+
+        sim_mom = sim_mom.reindex(index=emp_mom.index, columns=emp_mom.columns)
+        simulated_moments_reindexed.append(sim_mom)
+
+    return simulated_moments_reindexed
+
+
+def _replace_missing_values(simulated_moments, replace_nans):
+    """Replace missing values in simulated_moments."""
+    moments = []
+
+    for mom, func in zip(simulated_moments, replace_nans):
+        mom = func(mom)
+        moments.append(mom)
+
+    return moments
+
+
+def _flatten_index(data):
+    """Flatten the index as a combination of the former index and the columns."""
+    data = copy.deepcopy(data)
+    data_flat = []
+
+    for df in data:
```

Comment:
I am not sure whether this will work for ``pandas.Series`` as well. ``.unstack()`` will raise an ``AttributeError``.

After mapping ``str`` to the index, you need to a case distinction for DataFrames and Series and unstack only DataFrames.